File: src/main/java/cspfj/filter/AC3Constraint.java
package cspfj.filter;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Queue;

import scala.collection.IndexedSeq;
import scala.collection.JavaConversions;

import cspfj.ParameterManager;
import cspfj.StatisticsManager;
import cspfj.constraint.Constraint;
import cspfj.priorityqueues.BinomialHeap;
import cspfj.priorityqueues.Key;
import cspfj.problem.Problem;
import cspfj.problem.Variable;
import cspfj.util.Parameter;
import cspfj.util.Statistic;

/**
 * @author scand1sk
 * 
 */
public final class AC3Constraint implements Filter {

    private final Problem problem;

    private final Queue<Constraint> queue;

    // private static final Logger LOGGER = Logger.getLogger(Filter.class
    // .getSimpleName());

    @Statistic
    public int revisions = 0;

    private static int revisionCount = 0;

    @SuppressWarnings("rawtypes")
    @Parameter("ac.queue")
    private static Class<? extends Queue> queueType = BinomialHeap.class;

    @Parameter("ac.key")
    private static Key<Constraint> key = new Key<Constraint>() {
        @Override
        public double getKey(final Constraint object) {
            return object.getEvaluation();
        }

        @Override
        public String toString() {
            return "object.getEvaluation()";
        }
    };

    static {
        ParameterManager.register(AC3Constraint.class);
    }

    @SuppressWarnings("unchecked")
    public AC3Constraint(final Problem problem) {
        super();
        this.problem = problem;
        try {
            queue = queueType.getConstructor(Key.class).newInstance(key);
            StatisticsManager.register("ac.priorityQueue", queue);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    public boolean reduceAll() {
        revisionCount++;
        queue.clear();
        addAll();
        return reduce();

    }

    public boolean reduceFrom(final int[] modVar, final int[] modCons,
            final int cnt) {
        revisionCount++;
        queue.clear();
        // LOGGER.fine("reduce after " + cnt);
        for (Variable v : problem.getVariables()) {
            if (modVar[v.getId()] > cnt) {
                final IndexedSeq<Constraint> involved = v.constraints();
                for (int j = involved.size(); --j >= 0;) {
                    final Constraint constraint = involved.apply(j);
                    if (!constraint.isEntailed()) {
                        constraint.setRemovals(v.positionInConstraint()[j],
                                revisionCount);

                        queue.offer(involved.apply(j));
                    }
                }
            }
        }

        if (modCons != null) {
            for (Constraint c : JavaConversions.asJavaIterable(problem
                    .constraints())) {
                if (modCons[c.getId()] > cnt && !c.isEntailed()) {
                    c.fillRemovals(revisionCount);

                    queue.offer(c);
                }
            }
        }

        return reduce();
    }

    public boolean reduceAfter(final Variable variable) {
        revisionCount++;
        if (variable == null) {
            return true;
        }
        queue.clear();

        final IndexedSeq<Constraint> involving = variable.constraints();

        for (int cp = involving.size(); --cp >= 0;) {
            if (!involving.apply(cp).isEntailed()) {
                involving.apply(cp).setRemovals(
                        variable.positionInConstraint()[cp], revisionCount);
                queue.offer(involving.apply(cp));
            }
        }

        return reduce();
    }

    private RevisionHandler revisator = new RevisionHandler() {
        public void revised(final Constraint constraint, final Variable variable) {
            final IndexedSeq<Constraint> involvingConstraints = variable
                    .constraints();

            for (int cp = involvingConstraints.size(); --cp >= 0;) {
                final Constraint constraintP = involvingConstraints.apply(cp);
                if (constraintP != constraint && !constraintP.isEntailed()) {
                    constraintP.setRemovals(
                            variable.positionInConstraint()[cp], revisionCount);
                    queue.offer(constraintP);
                }

            }
        }
    };

    private boolean reduce() {
        // LOGGER.finer("Reducing");
        while (!queue.isEmpty()) {
            final Constraint constraint = queue.poll();

            revisions++;
            if (!constraint.revise(revisator, revisionCount)) {
                constraint.weight_$eq(constraint.weight() + 1);
                return false;
            }

            constraint.fillRemovals(-1);
        }

        assert control();

        return true;

    }

    private void addAll() {
        for (Constraint c : JavaConversions.asJavaIterable(problem
                .constraints())) {
            if (!c.isEntailed()) {
                c.fillRemovals(revisionCount);
                queue.offer(c);
            }
        }
    }

    private boolean control() {

        final RevisionHandler controlRevisator = new RevisionHandler() {

            @Override
            public void revised(final Constraint constraint,
                    final Variable variable) {
                assert false;

            }

        };

        for (Constraint c : JavaConversions.asJavaIterable(problem
                .constraints())) {
            assert c.revise(controlRevisator, -1);
        }
        return true;
    }

    public String toString() {
        return "GAC3rm-cons+" + queue.getClass().getSimpleName();
    }

    public Map<String, Object> getStatistics() {
        final Map<String, Object> statistics = new HashMap<String, Object>(1);
        statistics.put("revisions", revisions);
        return statistics;
    }

    @Override
    public boolean reduceAfter(final Collection<Constraint> constraints) {
        revisionCount++;
        queue.clear();

        for (Constraint c : constraints) {
            c.fillRemovals(revisionCount);
            queue.offer(c);
        }

        return reduce();
    }

}

File: src/main/scala/cspfj/filter/AC3Constraint.scala
package cspfj.filter;

import java.util.Queue
import scala.annotation.tailrec
import scala.collection.IndexedSeq
import cspfj.constraint.Constraint
import cspfj.priorityqueues._
import cspfj.problem.Problem
import cspfj.problem.Variable
import cspfj.util.Loggable
import cspfj.Parameter
import cspfj.ParameterManager
import cspfj.Statistic
import cspfj.StatisticsManager
import cspfj.constraint.Removals
import cspfj.problem.EmptyDomainException
import cspfj.heuristic.revision.Eval
import cspfj.UNSATException

object AC3Constraint {
  @Parameter("ac3c.queue")
  var queueType: Class[_ <: Queue[Constraint]] = classOf[BinomialHeap[Constraint]]

  @Parameter("ac3c.key")
  var keyType: Class[_ <: Key[Constraint]] = classOf[Eval]

  ParameterManager.register(this);

  def control(problem: Problem) = {

    for (c <- problem.constraints) {
      c.fillRemovals()
      val sizes = c.scope map (_.dom.size)
      assert(c.consistentRevise(), c + " is inconsistent");
      assert(
        sizes.sameElements(c.scope map (_.dom.size)),
        c + " was revised!")
    }

    true;
  }

  def key = keyType.getConstructor().newInstance()

  def queue = queueType.getConstructor(classOf[Key[Constraint]]).newInstance(key)
}

final class AC3Constraint(val problem: Problem, val queue: Queue[Constraint]) extends Filter with Loggable {
  @Statistic
  val substats = new StatisticsManager
  substats.register("ac.priorityQueue", queue);
  // private static final Logger LOGGER = Logger.getLogger(Filter.class
  // .getSimpleName());

  @Statistic
  var revisions = 0;

  def this(problem: Problem) = this(problem, AC3Constraint.queue)

  def reduceAll() = {
    Removals.clear()
    queue.clear()
    for (c <- problem.constraints if (!c.isEntailed)) {
      c.fillRemovals()
      queue.offer(c);
    }
    reduce()
  }

  def reduceFrom(modVar: Array[Int], modCons: Array[Int], cnt: Int) = {
    //Removals.clear()
    queue.clear();

    for (
      v <- problem.variables if (modVar(v.getId) > cnt)
    ) {
      updateQueue(v, null)
    }

    if (modCons != null)
      for (c <- problem.constraints if (modCons(c.getId) > cnt && !c.isEntailed)) {

        c.fillRemovals()
        queue.offer(c);

      }

    reduce();
  }

  def reduceAfter(variable: Variable) = variable == null || {
    Removals.clear()
    queue.clear();
    updateQueue(variable, null)
    reduce();
  }

  private def updateQueue(modified: Variable, skip: Constraint) {
    val constraints = modified.constraints

    @tailrec
    def upd(i: Int) {
      if (i >= 0) {
        val c = constraints(i)

        if ((c ne skip) && !c.isEntailed) {
          c.setRemovals(modified.positionInConstraint(i))
          queue.offer(c)
        }

        upd(i - 1)
      }
    }

    upd(constraints.size - 1)
  }

  @tailrec
  private def updateQueue(sizes: Array[Int], constraint: Constraint, i: Int) {
    if (i >= 0) {
      val v = constraint.scope(i)

      if (v.dom.size != sizes(i))
        updateQueue(v, constraint)

      updateQueue(sizes, constraint, i - 1)
    }
  }

  @tailrec
  private def reduce(): Boolean = {
    if (queue.isEmpty) {
      assert(AC3Constraint.control(problem))
      true
    } else {
      val constraint = queue.poll();

      revisions += 1;
      val sizes = constraint.sizes()

      val sat = try {
        if (constraint.revise()) {
          assert(!(constraint.sizes() sameElements sizes), constraint + " returned wrong true revised info")
          updateQueue(sizes, constraint, constraint.arity - 1)
        } else assert(constraint.sizes() sameElements sizes, constraint + " returned wrong false revised info")

        constraint.clearRemovals();
        true
      } catch {
        case e: UNSATException =>
          constraint.weight += 1
          false
      }
      sat && reduce()
    }
  }

  override def toString = "AC-cons+" + queue.getClass().getSimpleName();

  def getStatistics = Map("revisions" -> revisions)

  def reduceAfter(constraints: Iterable[Constraint]) = {
    Removals.clear()
    queue.clear();

    for (c <- constraints if !c.isEntailed) {
      c.fillRemovals()
      queue.offer(c);
    }

    reduce();
  }

}
